base_layer/core/src/base_node/comms_interface/inbound_handlers.rs

```diff
@@ -452,7 +447,22 @@ where B: BlockchainBackend + 'static
             );
             return Ok(());
         }
-
+        {
+            // we have this in mutex to ensure we dont have any funnies with rwlock with multiples requests first
+            // passing the check then all wanting to reconcile. We should only have 1 request passing
+            let mut lock = self.list_of_reconciling_blocks.lock().await;
```

Comment:
This could be made better by using a read/write lock and reading for the `contains` call, then obtaining a write for the insert call.

By adding the fast read, you avoid the attack of spamming a node with blocks and locking it up

Reply:
There may also be concurrent hashmaps that do this for you

Reply:
I looked at using a RWLock, but decided to use a Mutex because its safer and does what we want. 

Take this scenario:
Block 1-A -> read lock
Block 1-B -> read lock

Both has the read lock, both see that Block 1 does not exist. 

Both block 1-A and Block 1-B now get reconciled. And this should not happen.

Reply:
We can just make it write locks, that solves this, but then again we might just use a mutex